### PR TITLE
POC: Showcases for support type hints

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Related to https://github.com/GenericMappingTools/pygmt/issues/2794.

[Type Hints](https://docs.python.org/3/library/typing.html) are powerful and can be used to provide better auto-completion (address https://github.com/GenericMappingTools/pygmt/issues/1203). 

Here are some showcases in VSCode.

**Autocompletion for `**kwargs` parameters**

[autocompletion-parameters.webm](https://github.com/GenericMappingTools/pygmt/assets/3974108/ee5a71e3-7e9b-48f5-b445-52779f63d281)

**Autocompletion for parameter's allowed values**

https://github.com/GenericMappingTools/pygmt/assets/3974108/1371e553-4056-42c8-9dce-fe6a664c944c

**Autocompletion for dict keys**

Implement the syntax 1 in https://github.com/GenericMappingTools/pygmt/issues/1078#issuecomment-803246677 and address the concern in https://github.com/GenericMappingTools/pygmt/issues/1078#issuecomment-954361036

[autocompletion-dict-keys.webm](https://github.com/GenericMappingTools/pygmt/assets/3974108/0ef8c7ca-349e-4631-9fc0-c971c1047129)

### CODE REVIEW

File: pygmt/src/coast.py [hunk 1]
```diff
@@ -36,7 +37,26 @@
...
 )
 @kwargs_to_strings(R="sequence", c="sequence_comma", p="sequence")
-def coast(self, **kwargs):
+def coast(
+    self,
+    region: Any = None,
+    projection: Any = None,
+    area_thresh: Any = None,
+    frame: Any = None,
+    resolution: Literal["f", "h", "i", "l", "c", "a", None] = None,
```
[the remark this answers is not in the document]
Comment (seisman) [reply pygmt/src/coast.py:46]: I think it's worth a try.

File: pygmt/datasets/earth_relief.py [hunk 2]
```diff
@@ -13,11 +16,11 @@
...
 @kwargs_to_strings(region="sequence")
 def load_earth_relief(
-    resolution="01d",
-    region=None,
-    registration=None,
-    data_source="igpp",
-    use_srtm=False,
+    resolution: str = "01d",
+    region: Union[str, Sequence, None] = None,
+    registration: Literal["registration", "pixel", None] = None,
```
Comment (seisman) [inline pygmt/datasets/earth_relief.py:21]:
```suggestion
    registration: Literal["gridline", "pixel", None] = None,
```

**Comment by seisman** [comment]
Can anyone else check if changes in this PR work in Jupyter Notebooks? I can't make it work even after following this post (https://saturncloud.io/blog/autocomplete-jupyter-notebook/).

**Comment by weiji14** [comment]
> Can anyone else check if changes in this PR work in Jupyter Notebooks? I can't make it work even after following this post (https://saturncloud.io/blog/autocomplete-jupyter-notebook/).

Doesn't seem to work for me on JupyterLab 4.0.7:

Peek 2023-11-25 17-44

Using `%config IPCompleter.greedy=True` seems worse, it suggests `01d` :sweat_smile::

Peek 2023-11-25 17-49

**Pull request closed without merging**